internal/client.go

```diff
@@ -16,152 +16,207 @@ import (
 	"strings"
 
 	"github.com/PaesslerAG/jsonpath"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
+	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
+	"go.uber.org/zap"
 
 	"github.com/nutanix-cloud-native/prism-go-client"
 )
 
+type Scheme string
+
 const (
-	// libraryVersion = "v3"
-	defaultBaseURL = "%s://%s/"
-	httpPrefix     = "http"
-	httpsPrefix    = "https"
-	// absolutePath   = "api/nutanix/" + libraryVersion
-	// userAgent      = "nutanix/" + libraryVersion
+	defaultBaseURL  = "%s://%s/"
 	mediaType       = "application/json"
 	formEncodedType = "application/x-www-form-urlencoded"
 	octetStreamType = "application/octet-stream"
+
+	SchemeHTTP  Scheme = "http"
+	SchemeHTTPS Scheme = "https"
 )
 
-// Client Config Configuration of the client
+// RequestCompletionCallback defines the type of the request callback function
+type RequestCompletionCallback func(*http.Request, *http.Response, interface{})
+
+// Client Config Configuration of the httpClient
 type Client struct {
-	Credentials *prismgoclient.Credentials
+	credentials *prismgoclient.Credentials
 
-	// HTTP client used to communicate with the Nutanix API.
-	client *http.Client
+	// HTTP httpClient used to communicate with the Nutanix API.
+	httpClient *http.Client
 
 	// Base URL for API requests.
 	BaseURL *url.URL
 
-	// User agent for client
+	// User agent for httpClient
 	UserAgent string
 
-	Cookies []*http.Cookie
+	cookies []*http.Cookie
 
 	// Optional function called after every successful request made.
 	onRequestCompleted RequestCompletionCallback
 
 	// absolutePath: for example api/nutanix/v3
-	AbsolutePath string
+	absolutePath string
 
-	// error message, incase client is in error state
+	// error message, incase httpClient is in error state
 	ErrorMsg string
+
+	logger *logr.Logger
 }
 
-// RequestCompletionCallback defines the type of the request callback function
-type RequestCompletionCallback func(*http.Request, *http.Response, interface{})
+type ClientOption func(*Client) error
 
-// NewClient returns a wrapper around http/https (as per isHTTP flag) client with additions of proxy & session_auth if given
-func NewClient(credentials *prismgoclient.Credentials, userAgent string, absolutePath string, isHTTP bool) (*Client, error) {
-	if userAgent == "" {
-		return nil, fmt.Errorf("userAgent argument must be passed")
+// WithLogger sets the logger for the httpClient
+func WithLogger(logger *logr.Logger) ClientOption {
```

Comment:
Does it make sense to associate logger with all its variables to client as a whole? go-logr can attach Logger to context.Context. See https://pkg.go.dev/github.com/go-logr/logr#NewContext

Individual function extract logger from context then. See https://pkg.go.dev/github.com/go-logr/logr#FromContext

Logger is purely optional in context. A null logger is returned by FromContextOrDiscard() if not set.
Attaching logger to context becomes more necessary as call hierarchy gets deeper and each level attaches additional values to logger (in context).

Reply:
We can defer this until need arises.